oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,486 @@
+
+Maintaining Dependency Consistency
+----------------------------------
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity. That consistency doesn't come for free just by adopting shared dependencies at runtime: independent codebases will need to be kept in sync.
+
+We expect that this may need to take a number of possible forms.
+
+Process
+^^^^^^^
+
+We need to ensure maintainers and developers know what dependency versions to use, and when they need to upgrade to stay consistent. Open edX release documentation should include documentation of which frontend dependency versions are compatible with the release, likely pinned to a major version (i.e., React 17.x, Paragon 22.x, etc.)
+
+We also need a process to migrate core repositories through breaking changes in third-party dependencies. Ideally follow the `Upgrade Project Runbook <https://openedx.atlassian.net/wiki/spaces/AC/pages/3660316693/Upgrade+Project+Runbook>`_.
+
+Best Practices
+^^^^^^^^^^^^^^
+
+We need to ensure we minimize breaking changes in our own libraries (such as Paragon, the header, footer, frontend-platform, frontend-build, etc.) We suggest accomplishing this by:
+
+- Creating new versions of components with breaking changes (``ButtonV2``, ``webpack.dev.config.v2.js``) rather than modifying existing ones.
+- Leveraging the DEPR process for communication and removing old component versions
+- Aligning that removal and the subsequent breaking changes with Open edX releases, and documenting it in their release notes.
+
+Further, we could reduce the overhead of dependency maintenance and ensure MFEs stay up to date by pinning dependencies to major versions using ``^`` on versions in our package.json files.
+
+Tooling
+^^^^^^^
+
+Maintainers and developers should be warned of incompatibilities created by their PRs, or outside the repository by another project (such as the shell).
+
+This could take the form of Github tooling which notifies maintainers and developers that their frontend code has:
+
+1. Drifted behind the compatible version of a shared dependency for a given Open edX release or the main branch.
+2. Has upgraded beyond what is compatible with a given Open edX release or the main branch.
+
+Code Organization
+^^^^^^^^^^^^^^^^^
+
+We propose creating a passthrough library of shared dependencies. MFEs would depend on this single library in its package.json rather than on individual shared dependencies. This library would be versioned in accordance with the `Best Practices`_ suggestions above, meaning that breaking changes would be minimized and dependencies would be pinned to major versions via ``^`` on version numbers. Its version manifest would be the source of the version numbers for the `Process`_ and `Tooling`_ suggestions.
+
```

Comment:
Can you elobrate more on `passthrough library`  if possibe, I understood as:  _we are going to create monorepo that has header/footer/paragon...etc_. But I suspect that I am incorrect.

Reply:
We're not planning on creating a monorepo, this is a different idea to bundle our shared dependencies into a single library.  A monorepo is where the source code of multiple packages exists in the same repository. 

The idea of the 'passthrough' is that we'd have a library called something like `frontend-lib-shared-deps`.  It would basically do something like this:

```
export React from 'react';
export ReactDOM from 'react-dom';
export * as Paragon from '@openedx/paragon';
// etc.  The above may not be quite right, syntactically.
```

It'd basically take a bundle of dependencies and re-export them.  There's a lot of details glossed over in this - how do MFEs then import those modules?  Would we start doing `import { ReactDOM } from '@openedx/frontend-lib-shared-deps';` and stuff like that?  Do we introduce some webpack aliases so that we can continue to do `import ReactDOM from 'react-dom';`?  How does the versioning of this library work?  Do we use semantic versioning with it, and manually create releases that bundle a group of updates from dependencies so that we have to update less often?  How do we handle breaking changes in shared libraries?  I don't quite have answers to all that.  But the idea is to reduce the surface area of our dependencies to make it easier to stay up to date.  

I don't have a strong opinion on whether this OEP should have more _or less_ detail on these ideas.

Reply:
Maintaining this passthrough library sounds very complicated. Is the goal of reducing the number of breaking changes something we could solve with github tooling? We are already facing a similar issue with the maintenance of all the python libraries and we could reuse the same github tooling for both

Reply:
I found previous comments explaining more the need of the passthrough and even though I still find it complicated, I no longer think simple github tooling could solve it.

Reply:
Agreed about the tooling, unfortunately.  I'm adding some language here that explains why we might want to do this.